servo/runner.py

```diff
@@ -339,19 +345,25 @@ async def _report_progress(**kwargs) -> None:
                     f"failed progress reporting -- no current servo context is established (kwargs={devtools.pformat(kwargs)})"
                 )
 
-        def handle_progress_exception(error: Exception) -> None:
+        async def handle_progress_exception(progress: Dict[str, Any], error: Exception) -> None:
             # FIXME: This needs to be made multi-servo aware
             # Restart the main event loop if we get out of sync with the server
-            if isinstance(error, (servo.api.UnexpectedEventError, servo.api.EventCancelledError)):
-                if isinstance(error, servo.api.UnexpectedEventError):
+            if isinstance(error, (servo.errors.UnexpectedEventError, servo.errors.EventCancelledError)):
+                if isinstance(error, servo.errors.UnexpectedEventError):
                     self.logger.error(
                         "servo has lost synchronization with the optimizer: restarting"
                     )
-                elif isinstance(error, servo.api.EventCancelledError):
+                elif isinstance(error, servo.errors.EventCancelledError):
                     self.logger.error(
-                        "optimizer has cancelled operation in progress: restarting"
+                        "optimizer has cancelled operation in progress: cancelling and restarting loop"
                     )
 
+                    # Post a status to resolve the operation
+                    operation = progress['operation']
+                    status = servo.api.Status.from_error(error)
+                    runner = self._runner_for_servo(servo.current_servo())
+                    await runner._post_event(operation, status.dict())
+
                 tasks = [
                     t for t in asyncio.all_tasks() if t is not asyncio.current_task()
                 ]
```

Comment:
After [cancelling these tasks](https://github.com/opsani/servox/blob/fdfb83788e858fcf99e5e44f3b05f01a2eba3671/servo/runner.py#L371), they should be gathered with return_exceptions set to true to prevent the error message from unretrieved exceptions